Add network PMTU support

jukkar opened the pull request:
This implements IPv4 and IPv6 Path MTU Discovery support.
See https://www.rfc-editor.org/rfc/rfc1191 for IPv4 and
https://www.rfc-editor.org/rfc/rfc8201 for IPv6

General PMTU information can be found here https://en.wikipedia.org/wiki/Path_MTU_Discovery

Original issue here #2302

jukkar committed:
net: ip: pmtu: Add generic IP PMTU Discovery support

This adds generic code that can be used by both IPv4 and IPv6
Path MTU Discovery mechanism. The actual PMTU support for each
protocol family is in subsequent commits.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: pmtu: Add tests for path MTU discovery

Add initial tests for generic part of PMTU discovery code.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: stats: ipv6: pmtu: Add Path MTU Discovery statistics

Add information about PMTU related packets received/sent/dropped
for IPv6.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: ipv6: Print verdict information

Print more cases when the packet is dropped, and also print
the upper layer verdict for the packet.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: ipv6: Add PMTU support

Catch "Packet Too Big" ICMPv6 messages and update PMTU for
a given destination IPv6 address.
Use that PMTU when sending data to the destination.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: pmtu: Add IPv6 specific PMTU TCP tests

Allow tests to check whether a IPv6 TCP connection MTU is changed.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add pmtu command to the net-shell

Show information whether PMTU is enabled or not.
Show pmtu destination cache content with "net pmtu"
command. The "net pmtu flush" can be used to clear the
cache.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: stats: ipv4: pmtu: Add Path MTU Discovery statistics

Add information about PMTU related packets received/sent/dropped
for IPv4.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: ipv4: Add PMTU support

Catch "Destination Unreachable" ICMPv4 messages and update PMTU for
a given destination IPv4 address.
Use that PMTU when sending data to the destination.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: pmtu: Add IPv4 specific PMTU TCP tests

Allow tests to check whether a IPv4 TCP connection MTU is changed.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: pmtu: Send net_mgmt event for changed path MTU

Send a network management event for a changed path MTU value.
Both IPv4 and IPv6 have their own events as we cannot mix these
because how the network event numbering space is implemented.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: pmtu: Add network event tests for pmtu

Verify that the PMTU changed events are generated and we can
catch them.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: events: Print PMTU event values

Print the changed PMTU value and IP address in the event monitor.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: events: Set the command mask correctly

Set the commands to monitor correctly. Before this change some
events were missed.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
Note to self, the fragmentation code in both IPv4 and IPv6 needs some additional checks for PMTU. Need to fix that after 4.0 is released.

rlubos reviewed (dismissed):
Looks good, impressive work!

jukkar commented on include/zephyr/net/net_event.h:
```diff
@@ -441,6 +451,34 @@ struct net_event_ipv6_pe_filter {
+	/** IPv6 address */
+	struct in6_addr dst;
+	/** New MTU */
+	uint32_t mtu;
```
The IPv6 PMTU is specified as uint32_t in the RFC. We use only 16 bit value internally, but I thought that it might be good to use the length specified in the RFC, just in case we want to implement this internally too in the future. This way there is no need to change application facing interface.

jukkar commented:
Marking this as WIP/DNM as I need to implement these features:
* IPv4/IPv6 fragmentation code needs some tweaks for PMTU support
* There are socket options for setting/getting PMTU values from application. I will implement these in this PR.

jukkar committed:
net: Add support for IP_MTU IPv4 socket option

Add IP_MTU IPv4 socket option and implement getsockopt()
call for the option. The IP_MTU option does not support
setsockopt() call.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: pmtu: Add IP_MTU socket option tests

Make sure we can use IP_MTU socket option from application.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Add support for IPV6_MTU IPv6 socket option

Add IPV6_MTU IPv6 socket option and implement getsockopt() and
setsockopt() calls for the option.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: pmtu: Add IPV6_MTU socket option tests

Make sure we can use IPV6_MTU socket option from application.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: ipv4_fragment: Make test to run faster

Shorten the timeouts so that the tests are run in 6 second instead
of 15 seconds.

Use only native_sim for the tests so that the tests are run without
any extra delays.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: ipv4_fragment: Add PMTU support

If PMTU is enabled, then use the MTU value from it instead of always
using network interface MTU.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: ipv4_fragment: Add PMTU testing

Enable PMTU so that we test it with IPv4 fragmentation code.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: ipv6_fragment: Add PMTU support

If PMTU is enabled, then use the MTU value from it instead of always
using network interface MTU.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: ipv6_fragment: Add PMTU testing

Enable PMTU so that we test it with IPv6 fragmentation code.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* Changed the IPv4/6 fragmentation code to take PMTU into account
* Add `IPV6_MTU` and `IP_MTU` socket option support + tests for those

There is still `IP_MTU_DISCOVER` and `IPV6_MTU_DISCOVER` options that could be implemented, but that can be done in the future.

jukkar committed:
net: ipv6_fragment: Data in one frag must be multiple of 8

After we take the true MTU into account, we need to send proper
number of bytes (multiple of 8) in one IPv6 fragment.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* Rebased latest main
* Fixed the IPv6 fragment data count which must be a multiple of 8 bytes. The issue was not seen earlier because we always sent max. 1280 bytes in one fragment. But now the real MTU is used instead which need to check this.

rlubos commented on tests/net/ipv6_fragment/testcase.yaml:
```diff
@@ -1,8 +1,13 @@
 common:
```
nit; Typo in the commit message: `tests: net: ipvr6_fragment:` -> `tests: net: ipv6_fragment:`

jukkar commented:
@pdgendt if you have time, please take a look.

pdgendt commented on subsys/net/ip/pmtu.c:
```diff
@@ -0,0 +1,288 @@
+#define NET_IPV6_PMTU_ENTRIES 0
+#endif
+
+#define NET_PMTU_MAX_ENTRIES (NET_IPV4_PMTU_ENTRIES + NET_IPV6_PMTU_ENTRIES)
```
Minor hidden detail, the number of cache entries each get a maximum option setting, however AFAICS the entries are shared between IPv4 and IPv6?

So if both are set to 3, you could have 6 IPv4 and 0 IPv6 entries? Or am I missing something?
